parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoWriteSupport.java

```diff
@@ -247,21 +282,45 @@ final void writeRawValue(Object value) {
     @Override
     final void writeField(Object value) {
       recordConsumer.startField(fieldName, index);
+      recordConsumer.startGroup();
       List<?> list = (List<?>) value;
 
+      recordConsumer.startField("list", 0); // This is the wrapper group for the array field
       for (Object listEntry: list) {
+        recordConsumer.startGroup();
+
+        recordConsumer.startField("element", 0); // This is the mandatory inner field
+
+        if (!isPrimitive(listEntry)) {
+          recordConsumer.startGroup();
+        }
+
         fieldWriter.writeRawValue(listEntry);
+
+        if (!isPrimitive(listEntry)) {
+          recordConsumer.endGroup();
+        }
+
+        recordConsumer.endField("element", 0);
+
+        recordConsumer.endGroup();
       }
+      recordConsumer.endField("list", 0);
 
+      recordConsumer.endGroup();
       recordConsumer.endField(fieldName, index);
     }
   }
 
+  private boolean isPrimitive(Object listEntry) {
+    return !(listEntry instanceof Message);
```

Comment:
Could we do it without instanceof ?
We could pass the information to constructor of ArrayWriter

Reply:
done in https://github.com/costimuraru/parquet-mr/pull/2, I removed the isPrimitive method as wrapping with a group can be done in `MessageWriter.writeRawValue()`, removing the need to handle non-primitive types differently inside the ArrayWriter.